index.md

```diff
@@ -2,9 +2,343 @@
+
+Option #2 solves the collaboration problem at the expense of creating tens of thousands of buckets.
+Option #3 is similar to how we currently handle user data products but does potentially require more work to handle quotas than option #2.
+
+The choice between USDF and Cloud storage buckets depends on the cost of cloud storage versus USDF storage (based on the expected size of user data products) and how user batch {cite:p}`DMTN-223` interacts with cloud users.
+
+All user batch processing will take place at the USDF such that any user products to be used as inputs for that processing must be available at the USDF (the generated graph should not use signed URLs to locate the files in the cloud since signing happens in the server and the server is not available at batch execution time and signed URLs can not be stored in the graph because of expiry times) and all the data products created by the processing must be made available to the user and stored in their user or group collection.
+This constraint from batch processing suggests that USDF may be the best option since the files can be read from and written to the final location at USDF without requiring transfers.
+
+```{warning}
+It has been noted that Weka can only support up to 10,000 buckets.
+This is similar to the expected number of users and implies that if Weka is required as the USDF object store interface and user products are to be stored at USDF, then option #2 can not be used.
+```
+
+One additional thing to consider is how to handle user storage with multiple data releases.
+The first two options listed would need to have additional information in the file hierarchy to reference the relevant parent butler repository.
+The final option would use a server configuration to indicate where the specific user bucket is located as is done for DP0.2.
+
+#### Quotas
+
+```{note}
+Are users given distinct quota allocations for their RSP POSIX file space and Butler dataset usage?
+```
+
+Dataset file quotas for RSP users' POSIX file space will be monitored by the RSP infrastructure itself.
+Users will not have direct access to buckets without being mediated by the Butler server issuing signed URLs.
+
+For butler datasets quota management depends on how datasets are stored.
+For a shared bucket the butler server has to know (a) how much quota a user has, and (b) how much of the allocation they have utilized, before deciding whether a dataset can be stored.
+Care must be taken to deal with a user attempting multiple writes in parallel.
+
+Running a daily "cron" job to scan the bucket to calculate quota allocation does not seem feasible given how much people could store in a single day (especially for the outputs of user batch).
+This likely means that there will have to be some lightweight database tracking quotas for each user across all their buckets, with each write or deletion adjusting the quota value as appropriate.
```

Comment:
For systems like this, what has served me well in the past is to have a somewhat fast-and-loose system that adjusts on the fly but that may not be 100% correct in some situations, such as parallel writes, and then do a cron job comprehensive check on some interval to patch up the quota usage and reconcile any difference from reality.

Reply:
What I don't really know is whether reconciliation on a daily basis is fine and we don't really care if one user goes wildly over quota for a day before we notice.

Reply:
Hmm. @dhirving care to comment? If a single user is writing 500 puts in parallel and hitting 100 servers but one database backend, will locking in that one database block all the servers so that they end up doing serial writes and so can each get updated quota information.

Reply:
We wouldn't need to scan every object of every user every day, only the ones that may have been modified since the last check.

The main issues that may make it challenging to maintain an exact quota without a separate reconciliation step are:
1. It's possible for writes to fail due to network issues etc.  There's a sort of indeterminate state that exists after we issue a signed URL but before the URL expires or the client asks the server to finalize the put, where a file may or may not be in the process of being uploaded.
2. It may be inconvenient to require clients to declare the size of their files up front before we issue a signed URL
3. Users may have access to write into multiple Butler repositories which would share the quota
4. It's likely that the quota tracking will not be stored in the same DB as the registry itself, so we can't guarantee that the update is atomic.

Reply:
It seems difficult for issues 1 or 4 to be manipulated by the user to overuse the quota.  3 doesn't provide for enough parallelism to overuse the quota by much, if each repository on its own is nearly-consistent (it's only a small multiple of "nearly-").  For issue 2, if we do look at and deduct the size after the signed URL has been used and the client asks the server to finalize the put, the window for overuse still seems limited by the signed URL lifetime.
As I think was pointed out elsewhere, since the quota for cloud storage really is about cost, which is based on the integral of storage size over time rather than just max storage size, a limited lifetime basically completely mitigates the problem.